### PR TITLE
[Bugfix] Remove prequery properties from query_obj

### LINKED ISSUE #7666: Filter Box Caching Incorrectly for Multi-Query Use Case
A filter box that generates multiple queries (multiple filter controls) does not generate the correct cache entries for each filter control. It appears that only 1 (the first) query gets cached correctly. A correct cache entry should have a single query and associated results. What I see for most filter box cache entries are multiple queries from the filter box. So when the filter box gets reloaded, it only finds one of the cached queries and has cache misses for the rest. In each successive reload, the cache misses decrease by one. I suspect only the first query is generating the correct cache entry.

I am thinking the issue might be related to 'class FilterBoxViz' and/or 'run_extra_queries'
### Expected results

If I have a filter box with 16 filter controls, then a reload of the page within the cache TTL should generate 16 cache hits.

### Actual results

Only 1 cache hit and 15 cache misses. I see 15 more logs like 
2019-06-06 14:09:09,893:INFO:root:Caching 1180 chars at key c4dceda1ab0aef7047ef06ebb6063d88

#### Screenshots

Here is my filter box (16 controls)
Screen Shot 2019-06-06 at 9 59 17 AM

#### How to reproduce the bug
Set logging to INFO. I am using REDIS cache but probably any cache would do.
1. Load a dashboard containing a filter box with more than 1 filter control. Note occurrences of 
Caching xxxx chars at key yyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyy
2. Reload the page and note occurrences of "Caching abcd chars at key ..." and occurrences of "Serving from cache"
3. Note that there is only 1 cache hit for the filter box and that superset generated different cache keys then it did in step 1
4. Inspect the cache for the cache keys observed in step 1 and note that they contain multiple queries...they should contain a single query.

### Environment

(please complete the following information):

- superset version: 0.26.3
- python version: 2.7
- node.js version: not installed
- npm version: `not installed

### Checklist

Make sure these boxes are checked before submitting your issue - thank you!

- [x] I have checked the superset logs for python stacktraces and included it here as text if there are any.
- [x ] I have reproduced the issue with at least the latest released version of superset.
- [x] I have checked the issue tracker for the same issue and I haven't found one similar.

### PR DESCRIPTION
**Pull request by villebro**
### CATEGORY

Choose one

- [x] Bug Fix
- [x] Refactor

### SUMMARY

TL;DR:

There is a bug in how the `prequeries` property of `query_obj` is constructed in `sqla/models.py`, which causes cache misses if there are more than one filter selected in FilterBox. This PR fixes that bug without introducing any (user facing) functional changes.

Long explanation:

When non-join based prequeries were introduced in #4163, the properties `is_prequery` and `prequeries` were introduced into `query_obj`. This was done to be able to produce "Top N" filtering for engines that don't support joins, namely Druid (and later Pinot). However, prequeries are not actually needed in the `query_obj` (and later `cache_dict`), as they are only an intermediary construct that can be deduced from the `query_obj`. Furthermore, prequeries were  added to `query_obj` in `sqla_models.py`, making the calls non-idempotent and non-deterministic. Ideally `query_obj` should never be mutated after it is constructed in `viz.py`, as all the information necessary for the query should already be known at that time. This caused a caching bug in Filterbox, which runs as many queries as there are filters, and erroneously interpreted the previous queries as prequeries, causing bloated `prequeries` properties in the `query_obj`.

 When Pinot support was introduced in #6719, two named tuples (`SqlaQuery` and `QueryStrExtended`) were introduced in `sqla/models.py` to facilitate returning more context for how a query that produces the resulting `DataFrame` were produced. In the current codebase, this is the correct placeholder for prequeries. This PR removes both `is_prequery` and `prequeries` from `query_obj`, and adds `prequeries` into the `SqlaQuery` and `QueryStrExtended` objects, enabling the functionality introduced in #4163.

Functional changes:
- Removed `prequeries` and `is_prequery` from `query_obj`.
- Moved prequery related handling from `query_obj` to `SqlaQuery` and `QueryStrExtended`, ensuring that `query_obj` stays unmutated during processing in `sqla/models.py`.

Other refactoring:
- Renamed property `inner_joins` in `BaseEngineSpec` to `allows_joins` (currently not necessary to distinguish between inner/outer/full joins). Also harmonized other similarly named properties, such as `allows_subqueries` and `allows_column_aliases`.
- Added typing to `SqlaQuery`and `QueryStrExtended`.
- Moved some query formatting from `core.py` to `sqla/models.py`.

### TEST PLAN
- CI
- Tested locally by setting `allows_joins` to both `True` and `False` on `sqlite`
- Verified that no more cache misses occur when running Filterbox with more than one filter.

### ADDITIONAL INFORMATION
- [x] Has associated issue: Fixes #7666

### REVIEWERS
@smokemonster99

### CODE REVIEW

**Review by smokemonster99 — APPROVED** [review]
I tested in my dev environment on v33.0 and it looks good. LGTM

**Comment by villebro** [comment]
@smokemonster99 looking more closely this bug rquires delicate fixing, as the logic causing it is quite intricate. Will probably take a few days + a thorough round of review. In the meantime this WIP hack or your fix should be ok for your use case.

**Comment by villebro** [comment]
@smokemonster99 this is now ready; Can you test this in your env to make sure everything works as intended?

**Comment by mistercrunch** [comment]
@betodealmeida @agrawaldevesh

**Comment by villebro** [comment]
Yeah, this was some serious rabbit hole stuff 😅

**Comment by smokemonster99** [comment]
@villebro I will be traveling for meetings next week, I’ll see if I can get to it mid or late next week if you haven’t merged by then.

**Comment by villebro** [comment]
@smokemonster99 no prob and thanks for bringing this to our attention, good catch!

File: superset/connectors/sqla/models.py [hunk 1]
```diff
@@ -61,10 +61,18 @@
 config = app.config
 metadata = Model.metadata  # pylint: disable=no-member
 
-SqlaQuery = namedtuple(
-    "SqlaQuery", ["sqla_query", "labels_expected", "extra_cache_keys"]
-)
-QueryStringExtended = namedtuple("QueryStringExtended", ["sql", "labels_expected"])
+
+class SqlaQuery(NamedTuple):
```
Comment (mistercrunch) [inline superset/connectors/sqla/models.py]: Oh wow, the new `NamedTuple` are so much better, always hated the way the old ones were delcared...

**Review by mistercrunch — APPROVED** [review]
Did a pass and it LGTM, feel free to merge AFAIC

**Comment by villebro** [comment]
Thanks for the review @mistercrunch ; I think this turned out to be less invasive than I had originally thought, so will go ahead and merge.
